### DISCUSSION TITLE
Browse full server tree

### DISCUSSION THREAD

**Discussion by chialuc01** [FreeOpcUa/opcua-asyncio#1588] [Q&A] [answered]
Hello,
i have a client application which can retrieve all the nodes inside server tree in like 3 second. I'm imaging that is using a type of browse to retrieve in an async way all the data from the server. I've tried to implement your solution in the examples/client-minimal-auth.py but it's an iterative solution. Do you have something similar on your library?

**Comment by schroeder-** [comment]
First the is nothing you want build into the library. 
Also as opc ua works, you must do recursive browse calls. There is no shortcut for any client.

To speed things up, you can issue multiple browserequest in one call: https://reference.opcfoundation.org/Core/Part4/v104/docs/5.8.2
but then you have to use the lowlevel client and build the request our self.

Also you could try to use asyncio tasks, to do multiple calls at once.

**Reply by chialuc01** [reply]
Sorry but i have seen now that the client i was using is based on your project and it's called https://github.com/FreeOpcUa/opcua-client-gui . Is it possible to know how the application is able to load the tree on the left side in only 3 second for the whole server?

**Reply by schroeder-** [reply]
Don't know how client-gui works.
Think it browses the nodes on demand, thats why it is so quick, because it only browses the nodes that are shown/opend in the treeview.

**Reply by oroulet** [reply, marked as answer]
I wrote the client (a looong time ago) and yes it browses everytime you try to expand a node in the tree and uses the low level API at a few places.
Generally trying to load all the nodes is a very bad idea and probably impossible in under 3s depending on hw

**Reply by chialuc01** [reply]
Thank you! I understand the process now.
